File: android/domain/src/main/kotlin/dev/swiftshare/domain/PairingOrchestration.kt
package dev.swiftshare.domain

import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant

interface PairingInitiatorChannel {
    suspend fun send(message: PairingWireMessage)
    suspend fun receive(): PairingWireMessage
    suspend fun close()
}

fun interface PairingInitiatorConnecting {
    suspend fun connect(payload: PairingQrPayload): PairingInitiatorChannel
}

data class PairingInitiatorIdentity(
    val device: PairingWireDevice,
)

interface PairingInitiatorCryptography {
    suspend fun localIdentity(): PairingInitiatorIdentity
    suspend fun validateRemote(device: PairingWireDevice): PairingDeviceDescriptor
    suspend fun sign(digest: ByteArray): ByteArray
    suspend fun verify(device: PairingWireDevice, digest: ByteArray, signature: ByteArray): Boolean
}

fun interface PairingPeerCommitting {
    suspend fun commit(peer: PairingDeviceDescriptor, at: Instant)
}

data class PairingConfirmationRequest(
    val peer: PairingDeviceDescriptor,
    val comparisonCode: String,
)

fun interface PairingConfirming {
    suspend fun confirm(request: PairingConfirmationRequest): Boolean
}

sealed interface PairingInitiatorEvent {
    data object Connecting : PairingInitiatorEvent
    data class AwaitingConfirmation(val request: PairingConfirmationRequest) : PairingInitiatorEvent
    data object WaitingForRemoteCommit : PairingInitiatorEvent
    data class Completed(val peer: PairingDeviceDescriptor) : PairingInitiatorEvent
}

fun interface PairingInitiatorEventSink {
    suspend fun emit(event: PairingInitiatorEvent)
}

data class PairingExecutionOutcome(
    val peer: PairingDeviceDescriptor,
)

fun interface PairingExecutionClock {
    fun now(): Instant
}

fun interface PairingRandomGenerating {
    fun data(count: Int): ByteArray
}

class PairingInitiatorSession(
    private val connector: PairingInitiatorConnecting,
    private val cryptography: PairingInitiatorCryptography,
    private val committer: PairingPeerCommitting,
    private val confirmation: PairingConfirming,
    private val events: PairingInitiatorEventSink,
    private val clock: PairingExecutionClock = PairingExecutionClock(Instant::now),
    private val random: PairingRandomGenerating = PairingRandomGenerating {
        ByteArray(it).also(SecureRandom()::nextBytes)
    },
) {
    suspend fun execute(uri: String): PairingExecutionOutcome {
        val payload = try {
            PairingQrPayload.decode(uri, clock.now())
        } catch (error: Exception) {
            throw PairingException(PairingErrorCode.INVALID_QR)
        }
        events.emit(PairingInitiatorEvent.Connecting)
        val channel = connector.connect(payload)
        return try {
            run(payload, channel)
        } finally {
            channel.close()
        }
    }

    private suspend fun run(
        payload: PairingQrPayload,
        channel: PairingInitiatorChannel,
    ): PairingExecutionOutcome {
        val identity = cryptography.localIdentity()
        val localDescriptor = identity.device.descriptor
        val clientNonce = random.data(32)
        channel.send(ClientStart(payload.sessionId, payload.token, identity.device, clientNonce))

        val challenge = channel.receive() as? ServerChallenge
            ?: throw PairingException(PairingErrorCode.INVALID_TRANSCRIPT)
        val remoteDescriptor = cryptography.validateRemote(challenge.device)
        val transcript = try {
            PairingTranscript(
                payload.sessionId,
                payload.token,
                payload.certificateSha256,
                clientNonce,
                challenge.nonce,
                remoteDescriptor,
                localDescriptor,
            )
        } catch (error: Exception) {
            throw PairingException(PairingErrorCode.INVALID_TRANSCRIPT)
        }
        channel.send(ClientProof(cryptography.sign(transcript.proofDigest(PairingRole.ANDROID))))
        val proof = channel.receive() as? ServerProof
            ?: throw PairingException(PairingErrorCode.INVALID_TRANSCRIPT)
        if (!cryptography.verify(challenge.device, transcript.proofDigest(PairingRole.MAC), proof.signature)) {
            throw PairingException(PairingErrorCode.INVALID_PROOF)
        }

        val request = PairingConfirmationRequest(remoteDescriptor, transcript.comparisonCode)
        events.emit(PairingInitiatorEvent.AwaitingConfirmation(request))
        val accepted = confirmation.confirm(request)
        channel.send(
            ClientDecision(
                accepted,
                cryptography.sign(pairingDecisionDigest(transcript.digest, accepted)),
            ),
        )
        if (!accepted) throw PairingException(PairingErrorCode.CANCELLED)

        events.emit(PairingInitiatorEvent.WaitingForRemoteCommit)
        when (val response = channel.receive()) {
            is PairingWireError -> throw PairingException(
                if (response.code == "user_rejected") PairingErrorCode.CANCELLED else PairingErrorCode.INVALID_TRANSCRIPT,
            )
            is CommitReceipt -> {
                if (response.sessionId != payload.sessionId ||
                    !MessageDigest.isEqual(response.transcriptSha256, transcript.digest) ||
                    !MessageDigest.isEqual(response.macKeyId, remoteDescriptor.keyId) ||
                    !MessageDigest.isEqual(response.androidKeyId, localDescriptor.keyId)
                ) {
                    throw PairingException(PairingErrorCode.INVALID_TRANSCRIPT)
                }
                if (!cryptography.verify(
                        challenge.device,
                        pairingReceiptDigest(response),
                        response.signature,
                    )
                ) {
                    throw PairingException(PairingErrorCode.INVALID_PROOF)
                }
                try {
                    committer.commit(remoteDescriptor, clock.now())
                } catch (error: Exception) {
                    throw PairingException(PairingErrorCode.STORAGE)
                }
                val encoded = PairingWireCodec.encode(response)
                channel.send(CommitAck(MessageDigest.getInstance("SHA-256").digest(encoded)))
                events.emit(PairingInitiatorEvent.Completed(remoteDescriptor))
                return PairingExecutionOutcome(remoteDescriptor)
            }
            else -> throw PairingException(PairingErrorCode.INVALID_TRANSCRIPT)
        }
    }
}

fun pairingDecisionDigest(transcript: ByteArray, accepted: Boolean): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(
        "SwiftShare-Pairing-v1-decision-${if (accepted) "accept" else "reject"}".toByteArray() + transcript,
    )

fun pairingReceiptDigest(value: CommitReceipt): ByteArray = MessageDigest.getInstance("SHA-256").digest(
    "SwiftShare-Pairing-v1-receipt".toByteArray() +
        value.sessionId.toBytes() +
        value.transcriptSha256 +
        value.macKeyId +
        value.androidKeyId +
        ByteBuffer.allocate(8).putLong(value.committedAt.epochSecond).array(),
)

File: macos/SwiftShareDomain/PairingOrchestration.swift
import CryptoKit
import Foundation

public protocol PairingResponderChannel: Sendable {
    func send(_ message: PairingWireMessage) async throws
    func receive() async throws -> PairingWireMessage
    func close() async
}

public protocol PairingResponderListener: Sendable {
    func accept(payload: PairingQRPayload) async throws -> any PairingResponderChannel
    func close() async
}

public struct PairingResponderAdmission: Sendable {
    public let port: UInt16
    public let listener: any PairingResponderListener

    public init(port: UInt16, listener: any PairingResponderListener) {
        precondition(port > 0)
        self.port = port
        self.listener = listener
    }
}

public protocol PairingResponderListening: Sendable {
    func listen() async throws -> PairingResponderAdmission
}

public struct PairingResponderIdentity: Sendable {
    public let device: PairingWireDevice
    public let certificateSHA256: Data

    public init(device: PairingWireDevice, certificateSHA256: Data) {
        precondition(certificateSHA256.count == 32)
        self.device = device
        self.certificateSHA256 = certificateSHA256
    }
}

public protocol PairingResponderCryptography: Sendable {
    func localIdentity() async throws -> PairingResponderIdentity
    func validateRemote(_ device: PairingWireDevice) async throws -> PairingDeviceDescriptor
    func sign(_ digest: Data) async throws -> Data
    func verify(device: PairingWireDevice, digest: Data, signature: Data) async -> Bool
}

public protocol PairingPeerCommitting: Sendable {
    func commit(_ peer: PairingDeviceDescriptor, at date: Date) async throws
}

public struct PairingConfirmationRequest: Equatable, Sendable {
    public let peer: PairingDeviceDescriptor
    public let comparisonCode: String

    public init(peer: PairingDeviceDescriptor, comparisonCode: String) {
        self.peer = peer
        self.comparisonCode = comparisonCode
    }
}

public protocol PairingConfirming: Sendable {
    func confirm(_ request: PairingConfirmationRequest) async -> Bool
}

public enum PairingExecutionEvent: Equatable, Sendable {
    case qrIssued(PairingQRPayload)
    case waitingForRemoteCommit
    case completed(PairingDeviceDescriptor)
}

public protocol PairingExecutionEventSink: Sendable {
    func emit(_ event: PairingExecutionEvent) async throws
}

public struct PairingExecutionOutcome: Equatable, Sendable {
    public let peer: PairingDeviceDescriptor
    public init(peer: PairingDeviceDescriptor) { self.peer = peer }
}

public protocol PairingExecutionClock: Sendable {
    func now() -> Date
}

public protocol PairingRandomGenerating: Sendable {
    func uuid() -> UUID
    func data(count: Int) -> Data
}

public struct SystemPairingExecutionClock: PairingExecutionClock {
    public init() {}
    public func now() -> Date { Date() }
}

public struct SystemPairingRandomGenerator: PairingRandomGenerating {
    public init() {}
    public func uuid() -> UUID { UUID() }
    public func data(count: Int) -> Data {
        var generator = SystemRandomNumberGenerator()
        return Data((0 ..< count).map { _ in UInt8.random(in: .min ... .max, using: &generator) })
    }
}

public actor PairingResponderSession {
    private let listening: any PairingResponderListening
    private let cryptography: any PairingResponderCryptography
    private let committer: any PairingPeerCommitting
    private let confirmation: any PairingConfirming
    private let events: any PairingExecutionEventSink
    private let clock: any PairingExecutionClock
    private let random: any PairingRandomGenerating
    private let limits: PairingLimits

    public init(
        listening: any PairingResponderListening,
        cryptography: any PairingResponderCryptography,
        committer: any PairingPeerCommitting,
        confirmation: any PairingConfirming,
        events: any PairingExecutionEventSink,
        clock: any PairingExecutionClock = SystemPairingExecutionClock(),
        random: any PairingRandomGenerating = SystemPairingRandomGenerator(),
        limits: PairingLimits = PairingLimits()
    ) {
        self.listening = listening
        self.cryptography = cryptography
        self.committer = committer
        self.confirmation = confirmation
        self.events = events
        self.clock = clock
        self.random = random
        self.limits = limits
    }

    public func execute(host: String) async throws -> PairingExecutionOutcome {
        let identity = try await cryptography.localIdentity()
        let localDescriptor = try identity.device.descriptor
        let admission = try await listening.listen()
        let payload = try PairingQRPayload(
            sessionID: random.uuid(),
            host: host,
            port: admission.port,
            certificateSHA256: identity.certificateSHA256,
            token: random.data(count: 32),
            expiresAt: clock.now().addingTimeInterval(limits.admissionDuration)
        )
        let state = PairingSession(payload: payload, limits: limits)
        try await events.emit(.qrIssued(payload))

        let channel: any PairingResponderChannel
        do {
            channel = try await admission.listener.accept(payload: payload)
        } catch {
            await admission.listener.close()
            throw error
        }
        do {
            let outcome = try await run(
                channel: channel,
                payload: payload,
                state: state,
                identity: identity,
                localDescriptor: localDescriptor
            )
            await channel.close()
            await admission.listener.close()
            return outcome
        } catch {
            await channel.close()
            await admission.listener.close()
            throw error
        }
    }

    private func run(
        channel: any PairingResponderChannel,
        payload: PairingQRPayload,
        state: PairingSession,
        identity: PairingResponderIdentity,
        localDescriptor: PairingDeviceDescriptor
    ) async throws -> PairingExecutionOutcome {
        guard case .clientStart(let sessionID, let token, let remoteWire, let clientNonce) = try await channel.receive()
        else { throw PairingError.invalidTranscript }
        try await state.admit(sessionID: sessionID, token: token, now: clock.now())
        let remoteDescriptor = try await cryptography.validateRemote(remoteWire)
        let serverNonce = random.data(count: 32)
        try await channel.send(.serverChallenge(device: identity.device, nonce: serverNonce))
        let transcript = try PairingTranscript(
            sessionID: payload.sessionID,
            token: payload.token,
            certificateSHA256: payload.certificateSHA256,
            clientNonce: clientNonce,
            serverNonce: serverNonce,
            mac: localDescriptor,
            android: remoteDescriptor
        )

        guard case .clientProof(let remoteProof) = try await channel.receive(),
              await cryptography.verify(
                device: remoteWire,
                digest: transcript.proofDigest(for: .android),
                signature: remoteProof
              )
        else { throw PairingError.invalidProof }
        try await state.claim(candidateKeyID: remoteDescriptor.keyID, proofIsValid: true, now: clock.now())
        try await channel.send(.serverProof(signature: try await cryptography.sign(transcript.proofDigest(for: .mac))))

        let localAccepted = await confirmation.confirm(
            PairingConfirmationRequest(peer: remoteDescriptor, comparisonCode: transcript.comparisonCode)
        )
        guard case .clientDecision(let remoteAccepted, let remoteSignature) = try await channel.receive()
        else { throw PairingError.invalidTranscript }
        let remoteDecision = pairingDecisionDigest(transcript: transcript.digest, accepted: remoteAccepted)
        guard await cryptography.verify(device: remoteWire, digest: remoteDecision, signature: remoteSignature)
        else { throw PairingError.invalidProof }
        guard localAccepted, remoteAccepted else {
            await state.cancel()
            try await channel.send(.error(code: "user_rejected"))
            throw PairingError.cancelled
        }

        try await events.emit(.waitingForRemoteCommit)
        try await state.requireConfirmation(candidateKeyID: remoteDescriptor.keyID, now: clock.now())
        let committedAt = clock.now()
        do {
            try await committer.commit(remoteDescriptor, at: committedAt)
        } catch {
            throw PairingError.storage
        }
        try await state.commit(candidateKeyID: remoteDescriptor.keyID, now: committedAt)

        let receiptDigest = pairingReceiptDigest(
            sessionID: payload.sessionID,
            transcript: transcript.digest,
            mac: localDescriptor.keyID,
            android: remoteDescriptor.keyID,
            date: committedAt
        )
        let receipt = PairingWireMessage.commitReceipt(
            sessionID: payload.sessionID,
            transcriptSHA256: transcript.digest,
            macKeyID: localDescriptor.keyID,
            androidKeyID: remoteDescriptor.keyID,
            committedAt: committedAt,
            signature: try await cryptography.sign(receiptDigest)
        )
        try await channel.send(receipt)
        guard case .commitAck(let receivedDigest) = try await channel.receive(),
              receivedDigest == Data(SHA256.hash(data: PairingWireCodec.encode(receipt)))
        else { throw PairingError.invalidTranscript }

        try await events.emit(.completed(remoteDescriptor))
        return PairingExecutionOutcome(peer: remoteDescriptor)
    }
}

public func pairingDecisionDigest(transcript: Data, accepted: Bool) -> Data {
    Data(SHA256.hash(
        data: Data("SwiftShare-Pairing-v1-decision-\(accepted ? "accept" : "reject")".utf8) + transcript
    ))
}

public func pairingReceiptDigest(
    sessionID: UUID,
    transcript: Data,
    mac: Data,
    android: Data,
    date: Date
) -> Data {
    var seconds = UInt64(date.timeIntervalSince1970).bigEndian
    var uuid = sessionID.uuid
    return Data(SHA256.hash(
        data: Data("SwiftShare-Pairing-v1-receipt".utf8)
            + Data(bytes: &uuid, count: 16)
            + transcript
            + mac
            + android
            + Data(bytes: &seconds, count: 8)
    ))
}
